optimize/cmaes.go

```diff
@@ -0,0 +1,381 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package optimize
+
+import (
+	"math"
+	"math/rand"
+	"sort"
+	"sync"
+
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/mat"
+	"gonum.org/v1/gonum/stat/distmv"
+)
+
+// TODO(btracey): If we ever implement the traditional CMA-ES algorithm, provide
+// the base explanation there, and modify this description to just
+// describe the differences.
+
+// CmaEsChol implements the covariance matrix adaptation evolution strategy (CMA-ES)
+// based on the Cholesky decomposition. The full algorithm is described in
+//  Krause, Oswin, Dídac Rodríguez Arbonès, and Christian Igel. "CMA-ES with
+//  optimal covariance update and storage complexity." Advances in Neural
+//  Information Processing Systems. 2016.
+//  https://papers.nips.cc/paper/6457-cma-es-with-optimal-covariance-update-and-storage-complexity.pdf
+// CMA-ES is a global optimization method that progressively adapts a population
+// of samples. CMA-ES combines techniques from local optimization with global
+// optimization. Specifically, the CMA-ES algorithm uses an initial multivariate
+// normal distribution to generate a population of input locations. The input locations
+// with the lowest function values are used to update the parameters of the normal
+// distribution, a new set of input locations are generated, and this procedure
+// is iterated until convergence.
+//
+// As the normal distribution is progressively updated according to the best samples,
+// it can be that the mean of the distribution is updated in a gradient-descent
+// like fashion, followed by a shrinking covariance.
+// It is recommended that the algorithm be run multiple times (with different
+// InitMean) to have a better chance of finding the global minimum.
+//
+// The CMA-ES-Chol algorithm differs from the standard CMA-ES algorithm in that
+// it directly updates the Cholesky decomposition of the normal distribution.
+// This changes the runtime is O(dim^2*pop) rather than O(dim^3). The evolution
+// of the multi-variate normal will be similar to the baseline CMA-ES algorithm,
+// but the covariance update equation is not identical.
+//
+// For more information about the CMA-ES algorithm, see
+//  https://en.wikipedia.org/wiki/CMA-ES
+//  https://arxiv.org/pdf/1604.00772.pdf
+type CmaEsChol struct {
+	// InitStepSize sets the initial size of the covariance matrix adaptation.
+	// If InitStepSize is 0, a default vaule of 0.5 is used. InitStepSize cannot
+	// be negative.
```

Comment:
perhaps tell the user we will panic otherwise?
(perhaps I am being too pedantic)

Reply:
Done.